### PR TITLE
`unfold_backward` gets its own kernel

### LINKED ISSUE #17501: Tensor unfold backward is slow
## 🐛 Bug

Given the following code, the forward pass is very fast but the backward pass is unreasonably slow. I had a look at the backward pass code and while I am not an expert, I think it could be made significantly faster by writing a custom kernel instead of allocating index arrays, especially since `index_add_` seems to be so slow.

```python
import torch
import time

layer = torch.nn.Conv2d(3, 1, 3)
inp = torch.rand(10, 3, 1000, 1000)

out = layer(inp)
c = 5

res = out.unfold(3, c, 1)
res.sum().backward()

```

```
--------------------------------------------------------------------------------
  Environment Summary
--------------------------------------------------------------------------------
PyTorch 1.0.0.dev20181221 not compiled w/ CUDA
Running with Python 3.6 and

`pip list` truncated output:
Unable to fetch
--------------------------------------------------------------------------------
  cProfile output
--------------------------------------------------------------------------------
         1983 function calls (1400 primitive calls) in 8.831 seconds

   Ordered by: internal time
   List reduced from 79 to 15 due to restriction <15>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    8.216    8.216    8.216    8.216 {method 'run_backward' of 'torch._C._EngineBase' objects}
        1    0.315    0.315    0.315    0.315 {built-in method conv2d}
        1    0.218    0.218    0.218    0.218 {built-in method rand}
        1    0.078    0.078    0.078    0.078 {method 'sum' of 'torch._C._TensorBase' objects}
    102/1    0.001    0.000    0.002    0.002 /anaconda3/lib/python3.6/abc.py:196(__subclasscheck__)
        1    0.000    0.000    0.001    0.001 /anaconda3/lib/python3.6/site-packages/torch/nn/modules/conv.py:17(__init__)
        2    0.000    0.000    0.000    0.000 /anaconda3/lib/python3.6/site-packages/torch/nn/init.py:178(_calculate_fan_in_and_fan_out)
    98/30    0.000    0.000    0.001    0.000 /anaconda3/lib/python3.6/typing.py:1145(__subclasscheck__)
      202    0.000    0.000    0.000    0.000 /anaconda3/lib/python3.6/_weakrefset.py:70(__contains__)
        1    0.000    0.000    8.831    8.831 minimal.py:1(<module>)
        1    0.000    0.000    0.000    0.000 {method 'unfold' of 'torch._C._TensorBase' objects}
       85    0.000    0.000    0.000    0.000 /anaconda3/lib/python3.6/_weakrefset.py:58(__iter__)
    196/5    0.000    0.000    0.002    0.000 {built-in method builtins.issubclass}
       70    0.000    0.000    0.000    0.000 {method '__subclasses__' of 'type' objects}
       21    0.000    0.000    0.000    0.000 /anaconda3/lib/python3.6/site-packages/torch/nn/modules/module.py:537(__setattr__)


--------------------------------------------------------------------------------
  autograd profiler output (CPU mode)
--------------------------------------------------------------------------------
        top 15 events sorted by cpu_time_total

---------------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------
Name                                            CPU time        CUDA time            Calls        CPU total       CUDA total
---------------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------
UnfoldBackward                             7823869.000us          0.000us                1    7823869.000us          0.000us
index_add_                                 7222056.000us          0.000us                1    7222056.000us          0.000us
contiguous                                  370504.000us          0.000us                1     370504.000us          0.000us
clone                                       370501.000us          0.000us                1     370501.000us          0.000us
MkldnnConvolutionBackward                   257265.000us          0.000us                1     257265.000us          0.000us
mkldnn_convolution_backward                 257256.000us          0.000us                1     257256.000us          0.000us
mkldnn_convolution_backward_weights         257251.000us          0.000us                1     257251.000us          0.000us
conv2d                                      249499.000us          0.000us                1     249499.000us          0.000us
convolution                                 249495.000us          0.000us                1     249495.000us          0.000us
_convolution                                249494.000us          0.000us                1     249494.000us          0.000us
mkldnn_convolution                          249477.000us          0.000us                1     249477.000us          0.000us
contiguous                                  170065.000us          0.000us                1     170065.000us          0.000us
clone                                       170061.000us          0.000us                1     170061.000us          0.000us
sum                                          71872.000us          0.000us                1      71872.000us          0.000us
_th_arange                                   19876.000us          0.000us                1      19876.000us          0.000us
```

## To Reproduce

See code above

## Expected behavior

It should be faster

## Environment
Collecting environment information...
PyTorch version: 1.0.0.dev20181221
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.13.6
GCC version: Could not collect
CMake version: version 3.13.0

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] Could not collect
[conda] blas                      1.0                         mkl
[conda] mkl                       2018.0.3                      1
[conda] mkl-service               1.1.2            py37h6b9c3cc_5
[conda] mkl_fft                   1.0.6            py36hb8a8100_0
[conda] mkl_random                1.0.1            py36h5d10147_1
[conda] pytorch                   1.0.0                   py3.6_1    pytorch
[conda] pytorch-nightly           1.0.0.dev20181221         py3.6_0    pytorch
[conda] torchaudio                0.2                       <pip>
[conda] torchtext                 0.3.1                     <pip>
[conda] torchvision               0.2.1                      py_2    pytorch
## Additional context

My real application needs `c=200` but it takes more than 10 minutes to run (i stopped at that point). Let me know if the same operation can be done more efficiently with some other method.

cc @ezyang @gchanan @zou3519 @SsnL @albanD @gqchen @VitalyFedyunin @ngimel

### PR DESCRIPTION
**Pull request by nikitaved**
`unfold_backward` uses `index_add` which causes regression on CUDA because of the underlying `atomicAdd`, and regression on CPU because of limited parallelization. This PR attempts to replace `index_add` with a custom kernel.

Fixes [#17501](https://github.com/pytorch/pytorch/issues/17501).

### CODE REVIEW

**Comment by nikitaved** [comment]
@albanD , sorry to bother you, but could you please help me with the tests again? ;)
This implementation does not seem to be utterly wrong, and 10/12 autograd tests for unfold get passed. However, the two remaining `test_unfold_lastdim_*` fail. Could you tell me where I could find these? Grep, unfortunately, does not give me any results.

**Comment by albanD** [comment]
Sure!

These tests are automatically generated in `test_autograd.py` using the `add_test()` function.
The definition of everything that is passed to this function can be found [here](https://github.com/pytorch/pytorch/blob/46288465fe773da6941b2f366a0ad6e59f935d83/torch/testing/_internal/common_methods_invocations.py#L443-L445).

You can rerun these tests by doing `python test_autograd.py TestAutogradDeviceTypeCUDA.test_unfold_lastdim_neg0_cuda` for example. Even though you can't find this name in the file, this name is generated by the add_test() function.

**Comment by nikitaved** [comment]
Thank you! But objective does the autograd test the backward on?

**Comment by albanD** [comment]
In the file I link where you specify the test, you also specify the size of the Tensor inputs to the function and a random Tensor of this size is given.

**Comment by nikitaved** [comment]
Yes, but a need an objective to check the gradient. What is the underlying objective that gives me a scalar, not a tensor?

**Comment by albanD** [comment]
It computes a full Jacobian of that function.
So it will call one backward for each element in the output.

**Comment by nikitaved** [comment]
TODO: use `unsqueeze` on `grad_in`.

**Comment by nikitaved** [comment]
I am playing with a modification which is efficient for cases when `step >= size`.
However, during TensorIterator construction I am getting the following error during autograd testing:
```
RuntimeError: output with shape [1] doesn't match the broadcast shape [91, 3]
```
How come these shapes do not broadcast? The same holds when I use `[1,1]` shape.

The part of the code:
```
  auto grad_out_restrided = grad_out.unsqueeze(-1);
  auto grad_out_strides = ensure_nonempty_vec(grad_out.strides().vec());
  auto grad_out_sizes = ensure_nonempty_vec(grad_out.sizes().vec());
  grad_out_strides[dim] = 0;
  grad_out_sizes[dim] = 1;
  grad_out_restrided = grad_out_restrided.as_strided(grad_out_sizes, grad_out_strides);

  auto iter = TensorIterator();
  iter.dont_compute_common_dtype();
  iter.dont_resize_outputs();
  iter.add_output(grad_out_restrided);
  iter.add_input(grad_in, grad_in.device(), grad_in.scalar_type());
```
`grad_in` has an additional dimension of size `size`, and `grad_in.shape[dim] != grad_out.shape[dim]`, the rest is the same...

**Comment by nikitaved** [comment]
Ready for review.

**Comment by nikitaved** [comment]
Up to date benchmarks.
```python
import torch
from IPython import get_ipython

ipython = get_ipython()

#torch.set_num_threads(1)

layer = torch.nn.Conv2d(3, 1, 3)
inp = torch.rand(500, 3, 500, 500)

out = layer(inp)


for c in [3, 5, 7, 9, 11, 21]:
    res = out.unfold(3, c, 1)
    res_sum = res.contiguous().sum()

    print(f"c={c}")
    ipython.magic("timeit res_sum.backward(retain_graph=True)")

    del res

print()

for c in [3, 5, 7, 9, 11, 21]:
    res = out.unfold(3, c, c)
    res_sum = res.contiguous().sum()

    print(f"c={c}")
    ipython.magic("timeit res_sum.backward(retain_graph=True)")

    del res

```
### Master
#### CUDA
```
c=3
105 ms ± 220 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
c=5
119 ms ± 342 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
c=7
133 ms ± 325 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
c=9
147 ms ± 215 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
c=11
161 ms ± 363 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
c=21
231 ms ± 553 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

c=3
90.1 ms ± 209 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
c=5
93.9 ms ± 134 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
c=7
94 ms ± 240 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
c=9
94 ms ± 154 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
c=11
94 ms ± 130 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
c=21
94.2 ms ± 219 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
#### CPU
```
c=3
18.5 s ± 123 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
c=5
28 s ± 507 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
c=7
37.9 s ± 351 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
c=9
47.7 s ± 832 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
c=11
55.7 s ± 85.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
c=21
1min 42s ± 1.68 s per loop (mean ± std. dev. of 7 runs, 1 loop each)

c=3
8.64 s ± 75.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
c=5
8.67 s ± 70.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
c=7
9.4 s ± 436 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
c=9
9.18 s ± 212 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
c=11
9.52 s ± 155 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
c=21
9.46 s ± 325 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

```

### This PR
#### CUDA
```
c=3                                                                                                                                               
87 ms ± 545 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)                                                                               
c=5                                                                                                                                               
91.6 ms ± 229 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)                                                                             
c=7                                                                                                                                               
96.9 ms ± 253 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)                                                                             
c=9                                                                                                                                               
99.1 ms ± 392 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)                                                                             
c=11                                                                                                                                              
104 ms ± 175 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)                                                                              
c=21                                                                                                                                              
124 ms ± 278 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)                                                                              
                                                                                                                                                  
c=3                                                                                                                                               
85.2 ms ± 183 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)                                                                             
c=5                                                                                                                                               
85.4 ms ± 311 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)                                                                             
c=7                                                                                                                                               
85.8 ms ± 218 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
c=9
85.5 ms ± 222 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
c=11
85.8 ms ± 262 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
c=21
86 ms ± 163 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

```
#### CPU
```
c=3
4.64 s ± 349 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
c=5
4.38 s ± 155 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
c=7
4.34 s ± 115 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
c=9
4.42 s ± 79.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
c=11
4.5 s ± 93.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
c=21
4.99 s ± 79.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

c=3
4.1 s ± 168 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
c=5
4.08 s ± 105 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
c=7
4.16 s ± 159 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
c=9
4.22 s ± 164 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
c=11
4.2 s ± 139 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
c=21
4.07 s ± 169 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

File: aten/src/ATen/native/UnfoldBackward.h [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+  auto grad_in_strides = ensure_nonempty_vec(grad_in.strides().vec());
+  auto grad_in_sizes = ensure_nonempty_vec(grad_in.sizes().vec());
+
+  // set strides for dim to 0
+  // and size to 1 because
+  // this dimension is indexed inside the kernel
+  grad_in_strides[dim] = 0;
+  grad_in_sizes[dim] = 1;
+
+  grad_in_strides.pop_back();
```
Comment (albanD) [inline aten/src/ATen/native/UnfoldBackward.h:62]: Why do you remove the last dimension here?

File: aten/src/ATen/native/UnfoldBackward.h [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+  auto grad_out_dim = ensure_nonempty_dim(grad_out.dim());
+
+  auto idx_dim_strides = std::vector<int64_t>(grad_out_dim, 0);
+  auto idx_dim_sizes = std::vector<int64_t>(grad_out_dim, 1);
+
+  idx_dim_strides[dim] = 1;
+  idx_dim_sizes[dim] = iter_dim_size;
+
+  // idx_dim size will broadcast over determined by grad_out sizes in TensorIterator
+  auto idx_dim_restrided = idx_dim.as_strided(idx_dim_sizes, idx_dim_strides);
```
Comment (albanD) [inline aten/src/ATen/native/UnfoldBackward.h:87]: nit: wouldn't doing this with `.expand()` be simpler?

File: aten/src/ATen/native/cpu/UnfoldBackwardKernel.cpp [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  auto last_dim = maybe_wrap_dim(-1, grad_in.dim());
+
+  auto grad_in_dim_stride = ensure_nonempty_stride(grad_in, dim);
+  auto grad_in_last_dim_stride = ensure_nonempty_stride(grad_in, last_dim);
+  auto grad_in_dim_size = ensure_nonempty_size(grad_in, dim);
+
+  auto grad_out_dim_stride = ensure_nonempty_stride(grad_out, dim);
+
+  TensorIterator iter;
+  if (step >= size) {
```
Comment (albanD) [inline aten/src/ATen/native/cpu/UnfoldBackwardKernel.cpp:101]: nit: Given that you use this to make the choice between the two in at least two places, would it be clearer to use a boolean flag to choose this?

**Review by albanD — COMMENTED** [review]
I did not do a complete review but just looked at the big picture.
I think the architecture is good.
Still some build failures on CI to check.
It would be nice if you could do a quick high level description of the algorithm that you implement here to make the review of the iterators and kernels a bit easier.

File: aten/src/ATen/native/UnfoldBackward.h (hunk 1 above)
Comment (nikitaved) [reply aten/src/ATen/native/UnfoldBackward.h:62]: In this case I do not iterate over the last dimension at all. This iterator iterates over `grad_out`, which is of smaller dimension.

File: aten/src/ATen/native/UnfoldBackward.h (hunk 2 above)
Comment (nikitaved) [reply aten/src/ATen/native/UnfoldBackward.h:87]: Did not know about it, will check it out.

File: aten/src/ATen/native/cpu/UnfoldBackwardKernel.cpp (hunk 3 above)
Comment (nikitaved) [reply aten/src/ATen/native/cpu/UnfoldBackwardKernel.cpp:101]: I am sorry, I am not sure I understood you. Could you please elaborate?

**Comment by nikitaved** [comment]
The CI failures comes from not being able to understand `__restricted__`. I will try to fix that.

File: aten/src/ATen/native/cpu/UnfoldBackwardKernel.cpp (hunk 3 above)
Comment (albanD) [reply aten/src/ATen/native/cpu/UnfoldBackwardKernel.cpp:101]: Just that you have the same if inside the kernel itself to branch execution. Just felt a bit "dangerous". But this is a detail, nothing important :)

Comment (nikitaved) [reply aten/src/ATen/native/cpu/UnfoldBackwardKernel.cpp:101]: Got it! I agree. Will change to a boolean flag to reduce the potential 4 cases to only 2.

**Comment by nikitaved** [comment]
It seems like complex is not defined in device code. What shall I replace complex dispatch with?

File: aten/src/ATen/native/cuda/UnfoldBackwardKernel.cu [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    iter = _make_unfold_backward_iter_over_grad_out(
+      grad_out, grad_in, dim, size, step
+    );
+  }
+
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(
+    at::ScalarType::Half, at::ScalarType::Bool, at::ScalarType::BFloat16,
+    iter.dtype(),
+    "unfold_backward_cuda", [&] {
+      _unfold_backward_internal_kernel<scalar_t>(
```
Comment (anjali411) [inline aten/src/ATen/native/cuda/UnfoldBackwardKernel.cu:213]: std::complex is not supported on cuda, so we use thrust::complex instead. so I think you could do this:
```
using thrust_t = typename ztype_cuda<scalar_t>::thrust_t;
_unfold_backward_internal_kernel<thrust_t>(...)
``` similar to what's being done [here](https://github.com/pytorch/pytorch/blob/69e2f1aaff9614bd801b34b0c9b4cc8305ee9a61/aten/src/ATen/native/cuda/UnaryFractionKernels.cu)

cc. @zasdfgbnm

Comment (nikitaved) [reply aten/src/ATen/native/cuda/UnfoldBackwardKernel.cu]: Thanks, that worked!

**Review by albanD — COMMENTED** [review]
I only looked at the code structure and comment.
Will look at the algorithm next.

File: aten/src/ATen/native/cpu/UnfoldBackwardKernel.cpp [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+#endif
+
+// unfold_backward, the algorithm.
+//
+// Consider out = in.unfold(dim, size, step), then
+// out.shape[dim] == (in.shape[dim] - size) / step + 1,
+// out.shape[-1] == size.
+// out.dims() == in.dims() + 1
+//
+// unfold_backward receives grad_in and returns grad_out such that
```
[the remark this answers is not in the document]
Comment (nikitaved) [reply aten/src/ATen/native/cpu/UnfoldBackwardKernel.cpp:19]: Yeah, I actually use it the other way around. Shall I refactor that?

[the remark this answers is not in the document]
Comment (albanD) [reply aten/src/ATen/native/cpu/UnfoldBackwardKernel.cpp:19]: Refactor is a big word. Just renaming in the definition sounds enough to me !

[the remark this answers is not in the document]
Comment (nikitaved) [reply aten/src/ATen/native/cpu/UnfoldBackwardKernel.cpp:19]: The exposed function obeys the traditional naming convention, that's right. It is the kernel that flip things around. I will add a note about it then.

File: aten/src/ATen/native/cuda/UnfoldBackwardKernel.cu [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+      f(idx);
+      idx += n_threads;
+    }
+  }
+}
+
+template <int n_threads, int n_elems_per_thread, typename func_t>
+static void _launch_unfold_backward_kernel(int total_n_elems, func_t f) {
+  TORCH_INTERNAL_ASSERT(
+    total_n_elems >= 0 && total_n_elems <= std::numeric_limits<int32_t>::max()
```
Comment (albanD) [inline aten/src/ATen/native/cuda/UnfoldBackwardKernel.cu:39]: nit: `_unfold_backward_internal_kernel` already handle the case where total_n_elems==0. So you could change this to check `>0` and remove the if statement below I think.

**Review by albanD — APPROVED** [review]
Just a nit. Looks good!

File: aten/src/ATen/native/cuda/UnfoldBackwardKernel.cu (hunk 6 above)
Comment (nikitaved) [reply aten/src/ATen/native/cuda/UnfoldBackwardKernel.cu:39]: paying tribute to `legacy::launch_kernel` here ;)

**Pull request closed without merging**